File: vpo-backend/sound-engine/src/midi/messages.rs
use rhai::plugin::*;
use rhai::{Dynamic, Engine, EvalAltResult};

use serde::{Deserialize, Serialize};

pub type Channel = u8;
pub type Note = u8;
pub type Velocity = u8;
pub type Pressue = u8;
pub type ControlIndex = u8;
pub type ControlValue = u8;
pub type Patch = u8;
pub type Bend = i16;
pub type ExclusiveMessage = Vec<u8>;
pub type ManufacturerID = [u8; 3];
pub type TimecodeRate = u8;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Timecode {
    hours: u8,
    minutes: u8,
    seconds: u8,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "variant", content = "data")]
pub enum SystemCommonMessageData {
    SystemExclusive {
        id: ManufacturerID,
        message: ExclusiveMessage,
    },
    QuarterFrame {
        rate: TimecodeRate,
        time: Timecode,
    }, // Song Position Pointer
       // Song Select
       // Tune Request
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "variant", content = "data")]
pub enum SystemRealtimeMessageData {
    TimingClock,
    Start,
    Continue,
    Stop,
    ActiveSensing,
    Reset,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "variant", content = "data")]
pub enum MidiData {
    NoteOff {
        channel: Channel,
        note: Note,
        velocity: Velocity,
    },
    NoteOn {
        channel: Channel,
        note: Note,
        velocity: Velocity,
    },
    Aftertouch {
        channel: Channel,
        note: Note,
        pressure: Pressue,
    },
    ControlChange {
        channel: Channel,
        controller: ControlIndex,
        value: ControlValue,
    },
    ProgramChange {
        channel: Channel,
        patch: Patch,
    },
    ChannelAftertouch {
        channel: Channel,
        pressure: Pressue,
    },
    PitchBend {
        channel: Channel,
        pitch_bend: Bend,
    },
    SystemCommonMessage {
        data: SystemCommonMessageData,
    },
    SystemRealtimeMessage {
        data: SystemRealtimeMessageData,
    },
    MidiNone,
}

pub struct MidiMessage {
    pub data: MidiData,
    pub timestamp: u64,
}

File: vpo-frontend/src/sound-engine/midi/messages.ts
import type { DiscriminatedUnion } from "../../lib/util/discriminated-union";

export interface Timecode {
    hours: number,
    minutes: number,
    seconds: number,
}

export type SystemCommonMessageData = DiscriminatedUnion<"variant", {
    SystemExclusive: { content: { id: Uint8Array, message: Uint8Array }},
    QuarterFrame: { content: { rate: number, time: Timecode }},
}>;

export type SystemRealtimeMessageData = DiscriminatedUnion<"variant", {
    TimingClock: {},
    Start: {},
    Continue: {},
    Stop: {},
    ActiveSensing: {},
    Reset: {},
}>;

export type MidiData = DiscriminatedUnion<"variant", {
    NoteOff: { content: { channel: number, note: number, velocity: number }},
    NoteOn: { content: { channel: number, note: number, velocity: number }},
    Aftertouch: { content: { channel: number, note: number, pressure: number }},
    ControlChange: { content: { channel: number, controller: number, value: number }},
    ProgramChange: { content: { channel: number, patch: number }},
    ChannelAftertouch: { content: { channel: number, pitchBend: number }},
    PitchBend: { content: { channel: number, bend: number }},
    SystemCommonMessage: { content: { data: SystemCommonMessageData }},
    SystemRealtimeMessage: { content: { data: SystemRealtimeMessageData }},
    MidiNone: {}
}>;
